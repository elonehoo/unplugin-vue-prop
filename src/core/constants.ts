// Vue macros
export const DEFINE_PROPS = 'defineProps'
export const DEFINE_EMITS = 'defineEmits'
export const WITH_DEFAULTS = 'withDefaults'

export const PLUGIN_NAME = 'unplugin-vue-prop'

// Typescript types that the plugin allow to extract.
export const TS_TYPES_KEYS = ['TSTypeAliasDeclaration', 'TSInterfaceDeclaration', 'TSEnumDeclaration']
