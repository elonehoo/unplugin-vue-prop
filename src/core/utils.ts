import { basename, dirname, join, relative } from 'path'
import colors from 'picocolors'
import _debug from 'debug'
import fg from 'fast-glob'
import type { PackageInfo } from 'local-pkg'
import { getPackageInfoSync, resolveModule } from 'local-pkg'
import type { Alias, AliasOptions } from 'vite'
import { babelParse, generateCodeFrame } from '@vue/compiler-sfc'
import type { CallExpression, Node, Program, TSEnumDeclaration, TSInterfaceDeclaration, TSTypeAliasDeclaration } from '@babel/types'
import type { ExtractedTypes, IExport, IImport } from './ast'
import { DEFINE_EMITS, DEFINE_PROPS, PLUGIN_NAME, TS_TYPES_KEYS, WITH_DEFAULTS } from './constants'

/**
 * Type name prefixed with path
 *
 * @example '/foo/bar/baz.ts:Props'
 */
export type NameWithPath = string

/**
 * The actual name of type. It may be prefixed by the plugin.
 *
 * @example
 * ```text
 * 1. 'Foo'
 * 2. '_VTI_TYPE_Foo'
 * 3. '_VTI_TYPE_Foo_2'
 * ```
 */
export type FullName = string

export type TSTypes = TSTypeAliasDeclaration | TSInterfaceDeclaration | TSEnumDeclaration

export type MaybeAliases = ((AliasOptions | undefined) & Alias[]) | undefined

export type MaybeString = string | null | undefined

export type MaybeNumber = number | null | undefined

export type MaybeNode = Node | null | undefined

type Pkg = PackageInfo['packageJson']

interface PackageJSON extends Pkg {
  types?: string
  typings?: string
  exports?: {
    [p: string]: {
      types?: string
    }
  }
}

/**
 * References:
 * https://github.com/tc39/proposal-relative-indexing-method#polyfill
 * https://github.com/antfu/utils/blob/main/src/array.ts
 */
export function at(arr: [], index: number): undefined
export function at<T>(arr: T[], index: number): T
export function at<T>(arr: T[] | [], index: number): T | undefined {
  const length = arr.length

  if (index < 0)
    index += length

  if (index < 0 || index > length || !length)
    return undefined

  return arr[index]
}

export function debuggerFactory(namespace: string) {
  return (name?: string) => {
    const _debugger = _debug(`${PLUGIN_NAME}:${namespace}${name ? `:${name}` : ''}`)

    /**
     * NOTE(zorin): Use `console.log` instead when testing.
     * Because the output of the default logger is incomplete (i.e. it will lost some debug messages) when testing.
     */
    if (process.env.VITEST) {
      /* eslint-disable-next-line no-console */
      _debugger.log = console.log.bind(console)
    }

    return _debugger
  }
}

const createUtilsDebugger = debuggerFactory('Utils')

export function getAst(content: string): Program {
  return babelParse(content, {
    sourceType: 'module',
    plugins: ['typescript', 'topLevelAwait', 'jsx'],
  }).program
}

/**
 * Source: https://github.com/rollup/plugins/blob/master/packages/alias/src/index.ts
 */
export function matches(pattern: string | RegExp, importee: string) {
  if (pattern instanceof RegExp)
    return pattern.test(importee)

  if (importee.length < pattern.length)
    return false

  if (importee === pattern)
    return true

  const importeeStartsWithKey = importee.indexOf(pattern) === 0
  const importeeHasSlashAfterKey = importee.slice(pattern.length)[0] === '/'
  return importeeStartsWithKey && importeeHasSlashAfterKey
}

export function removeTSExtension(path: string): string {
  return path.replace(/\.ts$/, '')
}

export function resolvePath(path: string, from: string, aliases: MaybeAliases) {
  const debug = createUtilsDebugger('resolvePath')

  const matchedEntry = aliases?.find(entry => matches(entry.find, path))

  // Path which is using aliases. e.g. '~/types'
  if (matchedEntry)
    return path.replace(matchedEntry.find, matchedEntry.replacement)

  /**
   * External package
   * If the path is just a single dot, append '/index' to prevent incorrect results
   */
  const modulePath = resolveModule(path === '.' ? './index' : path)
  const dtsRE = /\.d\.ts$/

  // Not a package. e.g. '../types'
  if (!modulePath)
    return join(dirname(from), path)

  // Result is a typescript declaration file.
  if (dtsRE.test(modulePath)) {
    return removeTSExtension(modulePath)
  }
  // Not a typescript file, find declaration file
  else {
    const pkg = (getPackageInfoSync(path)?.packageJson || {}) as PackageJSON

    const slashArr = path.split('/')

    let processedPath = '.'

    // Increase index for scoped packages
    const index = slashArr[0][0] === '@' ? 2 : 1

    // Get relative path if module path contains slashes
    if (slashArr.length > index)
      processedPath = `./${relative(slashArr.slice(0, index).join('/'), path)}`

    debug('Processed path: %s', processedPath)

    const result: string = removeTSExtension(pkg.exports?.[processedPath]?.types || pkg.types || pkg.typings || basename(modulePath))

    return join(dirname(modulePath), result)
  }
}

export async function resolveModulePath(path: string, from: string, aliases?: MaybeAliases) {
  const debug = createUtilsDebugger('resolveModulePath')

  const maybePath = resolvePath(path, from, aliases)?.replace(/\\/g, '/')

  debug('Resolved path: %s', maybePath)

  if (!maybePath)
    return null

  // We follow the parsing order of typescript
  // https://www.typescriptlang.org/docs/handbook/module-resolution.html#how-typescript-resolves-modules
  // For modules: module > module/index
  // For extensions: .ts > .tsx > .d.ts
  const files = await fg([`${maybePath}.ts`, `${maybePath}.tsx`, `${maybePath}.d.ts`, `${maybePath}/index.ts`, `${maybePath}/index.tsx`, `${maybePath}/index.d.ts`], {
    onlyFiles: true,
  })

  debug('Matched files: %O', files)

  if (files.length)
    return files[0]

  return null
}

export type GroupedImports = Record<string, Record<string, string>>

export interface GroupedImportsResult {
  groupedImports: GroupedImports
  localSpecifierMap: Record<string, string>
}

/**
 * Categorize imports
 *
 * @example
 * ```
 * const code = `import { a as bb, b as cc, c as aa } from 'example'`
 * // ...(Operations of getting AST and imports)
 * const groupedImports = groupImports(imports);
 *
 * console.log(groupedImports)
 * // Result:
 * {
 *   groupedImports: {
 *     example: {
 *       bb: 'a',
 *       cc: 'b',
 *       aa: 'c'
 *     }
 *   },
 *   localSpecifierMap: {
 *     a: 'example',
 *     b: 'example',
 *     c: 'example'
 *   }
 * }
 * ```
 */
export function groupImports(imports: IImport[], source: string, fileName: string): GroupedImportsResult {
  const importedSpecifierMap: Record<string, string[]> = {}
  const localSpecifierMap: Record<string, string> = {}

  const groupedImports = imports.reduce<GroupedImports>((res, rawImport) => {
    const importedSpecifiers = importedSpecifierMap[rawImport.path]

    if (importedSpecifiers?.length && importedSpecifiers.includes(rawImport.imported)) {
      warn(`Duplicate imports of type "${rawImport.imported}" found.`, {
        fileName,
        codeFrame: generateCodeFrame(source, rawImport.start, rawImport.end),
      })
    }

    res[rawImport.path] ||= {}

    const aliases = res[rawImport.path]

    localSpecifierMap[rawImport.local] = rawImport.path

    importedSpecifierMap[rawImport.path] ||= []
    importedSpecifierMap[rawImport.path].push(rawImport.imported)

    if (rawImport.local !== rawImport.imported)
      aliases[rawImport.local] = rawImport.imported

    return res
  }, {})

  return {
    groupedImports,
    localSpecifierMap,
  }
}

/**
 * Convert export syntaxes to import syntaxes
 *
 * @example
 * Source:
 * ```typescript
 * export { Foo } from 'foo'
 * ```
 * Result:
 * ```typescript
 * import { Foo } from 'foo'
 * export { Foo }
 * ```
 */
export function convertExportsToImports(exports: IExport[], groupImportsResult: GroupedImportsResult): IImport[] {
  const { groupedImports, localSpecifierMap } = groupImportsResult

  return exports.map(({ start, end, local, exported, path }) => {
    if (path || localSpecifierMap[local]) {
      const mappedPath = localSpecifierMap[local]
      let imported = local

      if (!path && isString(mappedPath))
        imported = groupedImports[mappedPath][local] || local

      return {
        start,
        end,
        local: exported,
        imported,
        path: path || mappedPath,
      }
    }

    return null
  }).filter(notNullish)
}

// NOTE(zorin): Not used for now
export function intersect<A = any, B = any>(a: Array<A>, b: Array<B>): (A | B)[] {
  const setB = new Set(b)
  // @ts-expect-error unnecessary type checking (for now)
  return [...new Set(a)].filter(x => setB.has(x))
}

export interface Replacement {
  start: number
  end: number
  replacement: string
}

/**
 * Replace all items at specified indexes from the bottom up.
 *
 * NOTE(zorin): We assume that each replacement selection does not overlap with each other
 */
export function replaceAtIndexes(source: string, replacements: Replacement[], offset = 0): string {
  replacements.sort((a, b) => b.start - a.start)
  let result = source

  for (const node of replacements)
    result = result.slice(0, node.start + offset) + node.replacement + result.slice(node.end + offset)

  return result.split(/\r?\n/).filter(Boolean).join('\n')
}

/**
 * Collect dependencies and flatten it by using BFS
 *
 * NOTE(zorin): Maybe this needs a better solution.
 *
 * @example
 * Source code:
 * ```
 * type Foo = string;
 * type Bar = Foo;
 *
 * export interface Props {
 *   foo: Foo;
 *   bar: Bar;
 * }
 * ```
 * Dependency graph:
 * ```
 * Props ---Foo
 *       |--Bar--Foo
 * ```
 * Result: ['Foo', 'Bar', 'Props']
 */
export function resolveDependencies(extracted: ExtractedTypes, namesMap: Record<FullName, NameWithPath>, dependencies: string[]): string[] {
  function _resolveDependencies() {
    // NOTE(zorin): I don't think users will use same type for defineProps and defineEmits, so currently we do not dedupe them
    const queue: string[] = dependencies
    const result: string[] = []

    while (queue.length) {
      const shift = queue.shift()!

      const key = namesMap[shift]

      /**
       * Skip adding dependency
       *
       * NOTE(zorin): The only situation I know is invalid type (Types that are not even found after the recursion completes)
       */
      if (!key)
        continue

      result.push(key)

      const dependencies = extracted.get(key)!.dependencies

      if (dependencies?.length) {
        // Dedupe dependencies
        new Set(dependencies).forEach(dep => queue.push(dep))
      }
    }

    return result
  }

  // Dedupe from the end
  return [...new Set(_resolveDependencies().reverse())]
}

/**
 * Generate correct order of extension by using BFS. Similar to `resolveDependencies`
 *
 * NOTE(zorin): Maybe this needs a better solution.
 */
export function resolveExtends(record: Record<string, string[]>) {
  function _resolveExtends() {
    const queue: string[] = Object.keys(record)
    const result: string[] = []

    while (queue.length) {
      const shift = queue.shift()!

      result.push(shift)

      const extendTypes = record[shift]

      if (extendTypes?.length)
        extendTypes.forEach(key => queue.push(key))
    }

    return result
  }

  // Dedupe from the end
  const result = [...new Set(_resolveExtends().reverse())]

  return result
}

export function isNumber(n: MaybeNumber): n is number {
  return typeof n === 'number' && n.toString() !== 'NaN'
}

export function isString(n: MaybeString): n is string {
  return typeof n === 'string'
}

export function isCallOf(node: MaybeNode, test: string | ((id: string) => boolean)): node is CallExpression {
  return !!(
    node
    && node.type === 'CallExpression'
    && node.callee.type === 'Identifier'
    && (typeof test === 'string' ? node.callee.name === test : test(node.callee.name))
  )
}

export function isTSTypes(node: MaybeNode): node is TSTypes {
  return !!(node && TS_TYPES_KEYS.includes(node.type))
}

export function notNullish<T>(val: T | null | undefined): val is NonNullable<T> {
  return val != null
}

export interface LogOptions {
  fileName?: string
  codeFrame?: string
}

export function mergeLogMsg(options: LogOptions & { msg: string }) {
  const { msg, fileName, codeFrame } = options

  // NOTE(zorin): We only log basic message in test
  const result = [
    msg,
    '',
    process.env.VITEST ? undefined : fileName,
    process.env.VITEST ? undefined : codeFrame,
  ].filter(notNullish)

  // Push newline if the last line is not an empty string
  if (at(result, -1))
    result.push('')

  return result.join('\n')
}

export function warn(msg: string, { fileName, codeFrame }: LogOptions = {}) {
  const result = mergeLogMsg({ msg, fileName, codeFrame })

  console.warn(colors.yellow(`[${PLUGIN_NAME}] WARN: ${result}`))
}

export function error(msg: string, { fileName, codeFrame }: LogOptions = {}): never {
  const result = mergeLogMsg({ msg, fileName, codeFrame })

  throw new Error(colors.red(`[${PLUGIN_NAME}] ERROR: ${result}`))
}

export function isEnum(node: TSTypes | null | undefined): node is TSEnumDeclaration {
  return !!(node && node.type === 'TSEnumDeclaration')
}

export const isDefineProps = (node: Node): node is CallExpression => isCallOf(node, DEFINE_PROPS)
export const isDefineEmits = (node: Node): node is CallExpression => isCallOf(node, DEFINE_EMITS)
export const isWithDefaults = (node: Node): node is CallExpression => isCallOf(node, WITH_DEFAULTS)
