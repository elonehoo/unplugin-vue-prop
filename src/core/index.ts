import { parse } from '@vue/compiler-sfc'
import MagicString from 'magic-string'
import type { TransformResult } from 'rollup'
import type { ExtractResult, TypeMetaData } from './ast'
import { extractTypesFromSource, getUsedInterfacesFromAst } from './ast'
import { debuggerFactory, getAst, notNullish, replaceAtIndexes, resolveDependencies, resolveExtends } from './utils'
import type { MaybeAliases, Replacement } from './utils'

export interface CleanOptions {
  interface?: boolean
}

export interface TransformOptions {
  id: string
  aliases?: MaybeAliases
}

export interface FinalizeResult {
  inlinedTypes: string
  replacements: Replacement[]
}

const createMainDebugger = debuggerFactory('Main')

export function finalize(types: string[], extractResult: ExtractResult): FinalizeResult | null {
  const debug = createMainDebugger('Finalize')

  const { result, namesMap, typeReplacements, extendsRecord, importNodes, extraSpecifiers, sourceReplacements } = extractResult

  if (!result.size)
    return null

  Object.entries(typeReplacements).forEach(([k, r]) => {
    debug('Replacements %s => %O', k, r)
  })

  debug('Keys map: %O', namesMap)

  // Apply replacements
  Object.entries(typeReplacements).forEach(([key, replacementInfo]) => {
    const extractedTypeInfo = result.get(key)!
    const { offset, replacements } = replacementInfo

    extractedTypeInfo.body = replaceAtIndexes(extractedTypeInfo.body, replacements, offset)
  })

  const resolvedExtendsOrder = resolveExtends(extendsRecord)

  debug('Resolved extends order: %O', resolvedExtendsOrder)

  // Insert code for extended interfaces in order
  resolvedExtendsOrder.forEach((key) => {
    const extractedTypeInfo = result.get(key)!

    const extendTypes = extendsRecord[key]

    if (!extendTypes?.length)
      return

    const replacements: Replacement[] = extendTypes.map((key) => {
      const { body, dependencies } = result.get(key)!

      // Add dependencies for extended interfaces
      if (dependencies?.length) {
        extractedTypeInfo.dependencies ||= []
        extractedTypeInfo.dependencies!.push(...dependencies)
      }

      return {
        start: 1,
        end: 1,
        replacement: body.slice(1, body.length - 1),
      }
    })

    extractedTypeInfo.body = replaceAtIndexes(extractedTypeInfo.body, replacements)
  })

  debug('Result: %O', result)
  debug('Extra specifiers 3: %O', extraSpecifiers)

  // Collect replacements to clean up import specifiers
  importNodes.forEach((i) => {
    let defaultSpecifier: string | undefined

    const savedSpecifiers = i.specifiers
      .map((specifier) => {
        if (specifier.type === 'ImportSpecifier' && specifier.imported.type === 'Identifier' && specifier.local.type === 'Identifier') {
          const imported = specifier.imported.name
          const local = specifier.local.name

          let fullName = local

          /**
           * NOTE(zorin): We only remove specifiers that used directly by user because the name of their dependencies are prefixed by the plugin
           */
          const shouldSave = !types.includes(local)

          if (shouldSave && !extraSpecifiers.includes(local)) {
            if (imported !== local)
              fullName = `${imported} as ${local}`

            return fullName
          }

          return null
        }
        else if (specifier.type === 'ImportDefaultSpecifier') {
          const name = specifier.local.name

          if (!types.includes(name) && !extraSpecifiers.includes(name))
            defaultSpecifier = name
        }

        return null
      })
      .filter(notNullish)

    const replacement: string[] = [
      'import',
    ]

    if (defaultSpecifier)
      replacement.push(` ${defaultSpecifier}`)

    if (savedSpecifiers.length)
      replacement.push(`${defaultSpecifier ? ',' : ''} { ${savedSpecifiers.join(', ')} }`)

    // Remove the import statement if no specifiers are saved
    if (replacement.length === 1) {
      sourceReplacements.push({
        start: i.start!,
        end: i.end!,
        replacement: '',
      })
    }
    // Generate a new import statement to replace the original one
    else {
      replacement.push(` from '${i.source.value}'`)

      sourceReplacements.push({
        start: i.start!,
        end: i.end!,
        replacement: replacement.join(''),
      })
    }
  })

  /**
   * NOTE(zorin): Get the order of inlining types
   * If the order is incorrect, Vue will not get the correct definition of types
   *
   * @example
   * Correct:
   * ```typescript
   * type Foo = string
   * type Bar = Foo
   *
   * interface Props {...}
   * ```
   * Wrong:
   * ```typescript
   * type Bar = Foo
   * type Foo = string
   *
   * interface Props {...}
   * ```
   */
  const dependencies = resolveDependencies(result, namesMap, types)

  debug('Dependencies: %O', dependencies)

  const inlinedTypes = dependencies.map((key) => {
    const { typeKeyword, fullName, body } = result.get(key)!

    let maybeEqualSign = ''

    if (typeKeyword === 'type')
      maybeEqualSign = ' ='

    return `${typeKeyword} ${fullName}${maybeEqualSign} ${body}`
  }).join('\n')

  return {
    inlinedTypes,
    replacements: sourceReplacements,
  }
}

export async function transform(code: string, { id, aliases }: TransformOptions): Promise<TransformResult | void> {
  const {
    descriptor: { scriptSetup },
  } = parse(code)

  const isTS = scriptSetup && (scriptSetup.lang === 'ts' || scriptSetup.lang === 'tsx')

  if (!isTS || !scriptSetup.content){
    return
  }
    

  const program = getAst(scriptSetup.content)

  const interfaces = getUsedInterfacesFromAst(program)

  const metaDataMap = Object.fromEntries(interfaces.map<[string, TypeMetaData]>(name => [name, { isUsedType: true }]))

  const extractResult = await extractTypesFromSource(
    scriptSetup.content,
    interfaces,
    {
      pathAliases: aliases,
      relativePath: id,
      metaDataMap,
      ast: program,
      isInSFC: true,
    },
  )

  const result = finalize(interfaces, extractResult)

  if (!result){
    return
  }
    

  const { inlinedTypes, replacements } = result

  const s = new MagicString(code)

  // Add inlined types and replace import statements
  const newScriptSetupContent = [
    '',
    inlinedTypes,
    replaceAtIndexes(scriptSetup.content, replacements),
    '',
  ].join('\n')

  s.overwrite(scriptSetup.loc.start.offset, scriptSetup.loc.end.offset, newScriptSetupContent)

  if (s.hasChanged()) {
    return {
      code: s.toString(),
      map:s.generateMap({
        source: id,
        includeContent: true,
        hires: true,
      })
    }
  }
}
