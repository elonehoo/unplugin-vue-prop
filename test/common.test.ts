import { removeTSExtension } from '../src/core/utils'
import type { CodeGetter } from './_utils'
import { defineTransformTest } from './_utils'

const codeGetter: CodeGetter = ({ entry }) => `<script lang="ts" setup>
import { Props } from '${removeTSExtension(entry)}'

defineProps<Props>()
</script>
`

defineTransformTest({
  category: 'Common',
  filePattern: ['./fixtures/common/**/!(_)*.ts'],
  fileName: __filename,
  codeGetter,
  skip: false,
})
