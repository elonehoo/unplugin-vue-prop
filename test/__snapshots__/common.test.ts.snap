// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Common > Duplicate imports > Export aliases > '1.ts' ('default') 1`] = `
"<script lang=\\"ts\\" setup>
type _VTI_TYPE_Bar = 'foo'
interface Props {
  foo: _VTI_TYPE_Bar
  bar: _VTI_TYPE_Bar
}
defineProps<Props>()
</script>
"
`;

exports[`Common > Duplicate imports > Export aliases > '2.ts' ('default') 1`] = `
"<script lang=\\"ts\\" setup>
type _VTI_TYPE_Foo = 'foo'
interface Props {
  foo: _VTI_TYPE_Foo
  bar: _VTI_TYPE_Foo
}
defineProps<Props>()
</script>
"
`;

exports[`Common > Duplicate imports > Export aliases > 'multi-level.ts' ('default') 1`] = `
"<script lang=\\"ts\\" setup>
type _VTI_TYPE_Foo = 'foo'
interface Props {
  foo: _VTI_TYPE_Foo
  bar: _VTI_TYPE_Foo
}
defineProps<Props>()
</script>
"
`;

exports[`Common > Duplicate imports > Import aliases > '1.ts' ('default') 1`] = `
"<script lang=\\"ts\\" setup>
type _VTI_TYPE_Bar = 'foo'
interface Props {
  foo: _VTI_TYPE_Bar
  bar: _VTI_TYPE_Bar
}
defineProps<Props>()
</script>
"
`;

exports[`Common > Duplicate imports > Import aliases > '2.ts' ('default') 1`] = `
"<script lang=\\"ts\\" setup>
type _VTI_TYPE_Bar = 'foo'
interface Props {
  foo: _VTI_TYPE_Bar
  bar: _VTI_TYPE_Bar
}
defineProps<Props>()
</script>
"
`;

exports[`Common > Duplicate imports > Import aliases > 'multi-level.ts' ('default') 1`] = `
"<script lang=\\"ts\\" setup>
type _VTI_TYPE_Bar = 'foo'
interface Props {
  foo: _VTI_TYPE_Bar
  bar: _VTI_TYPE_Bar
}
defineProps<Props>()
</script>
"
`;

exports[`Common > Duplicate imports > Import export default > '1.ts' ('default') 1`] = `
"<script lang=\\"ts\\" setup>
type _VTI_TYPE_Foo = 'foo'
interface Props {
  foo: _VTI_TYPE_Foo
  bar: _VTI_TYPE_Foo
}
defineProps<Props>()
</script>
"
`;

exports[`Common > Duplicate imports > Import export default > '2.ts' ('default') 1`] = `
"<script lang=\\"ts\\" setup>
type _VTI_TYPE_Foo = 'foo'
interface Props {
  foo: _VTI_TYPE_Foo
  bar: _VTI_TYPE_Foo
}
defineProps<Props>()
</script>
"
`;

exports[`Common > Duplicate imports > Import export default > 'multi-level.ts' ('default') 1`] = `
"<script lang=\\"ts\\" setup>
type _VTI_TYPE_Foo = 'foo'
interface Props {
  foo: _VTI_TYPE_Foo
  bar: _VTI_TYPE_Foo
}
defineProps<Props>()
</script>
"
`;

exports[`Common > Enum types > Default > 'empty.ts' ('default') 1`] = `
"<script lang=\\"ts\\" setup>
type _VTI_TYPE_Foo = /* empty-enum */ number | string
interface Props {
  foo: _VTI_TYPE_Foo
}
defineProps<Props>()
</script>
"
`;

exports[`Common > Enum types > Default > 'mixed.ts' ('default') 1`] = `
"<script lang=\\"ts\\" setup>
type _VTI_TYPE_Foo = /* enum */ number | string
interface Props {
  foo: _VTI_TYPE_Foo
}
defineProps<Props>()
</script>
"
`;

exports[`Common > Enum types > Default > 'number.ts' ('default') 1`] = `
"<script lang=\\"ts\\" setup>
type _VTI_TYPE_Foo = /* enum */ number
interface Props {
  foo: _VTI_TYPE_Foo
}
defineProps<Props>()
</script>
"
`;

exports[`Common > Enum types > Default > 'string.ts' ('default') 1`] = `
"<script lang=\\"ts\\" setup>
type _VTI_TYPE_Foo = /* enum */ string
interface Props {
  foo: _VTI_TYPE_Foo
}
defineProps<Props>()
</script>
"
`;

exports[`Common > Export aliases > Default > 'index.ts' ('default') 1`] = `
"<script lang=\\"ts\\" setup>
type _VTI_TYPE_Foo = string
interface Props {
  foo: _VTI_TYPE_Foo
}
defineProps<Props>()
</script>
"
`;

exports[`Common > Export aliases > Multi level > '1.ts' ('default') 1`] = `
"<script lang=\\"ts\\" setup>
interface Props {
  foo: string
}
defineProps<Props>()
</script>
"
`;

exports[`Common > Export aliases > Multi level > '2.ts' ('default') 1`] = `
"<script lang=\\"ts\\" setup>
interface Props {
  foo: string
}
defineProps<Props>()
</script>
"
`;

exports[`Common > Export all > Default > 'index.ts' ('default') 1`] = `
"<script lang=\\"ts\\" setup>
type _VTI_TYPE_Foo = 'foo'
interface Props {
  foo: _VTI_TYPE_Foo
}
defineProps<Props>()
</script>
"
`;

exports[`Common > Import aliases > Default > 'index.ts' ('default') 1`] = `
"<script lang=\\"ts\\" setup>
type _VTI_TYPE_Foo = number
interface Props {
  foo: _VTI_TYPE_Foo
}
defineProps<Props>()
</script>
"
`;

exports[`Common > Import aliases > Multi level > 'index.ts' ('default') 1`] = `
"<script lang=\\"ts\\" setup>
type _VTI_TYPE_Bar = string
type _VTI_TYPE_Foo = number | _VTI_TYPE_Bar
interface Props {
  foo: _VTI_TYPE_Foo
}
defineProps<Props>()
</script>
"
`;

exports[`Common > Import export default > Default > '1.ts' ('default') 1`] = `
"<script lang=\\"ts\\" setup>
type _VTI_TYPE_Foo = string
interface Props {
  foo: _VTI_TYPE_Foo
}
defineProps<Props>()
</script>
"
`;

exports[`Common > Import export default > Multi level > '1.ts' ('default') 1`] = `
"<script lang=\\"ts\\" setup>
interface Props {
  foo: number
}
defineProps<Props>()
</script>
"
`;

exports[`Common > Import export default > Multi level > '2.ts' ('default') 1`] = `
"<script lang=\\"ts\\" setup>
type _VTI_TYPE_Foo = string
interface Props {
  foo: _VTI_TYPE_Foo
}
defineProps<Props>()
</script>
"
`;

exports[`Common > Import export default > Use aliases > '1.ts' ('default') 1`] = `
"<script lang=\\"ts\\" setup>
type _VTI_TYPE_Foo = string
interface Props {
  foo: _VTI_TYPE_Foo
}
defineProps<Props>()
</script>
"
`;

exports[`Common > Import export default > Use aliases > '2.ts' ('default') 1`] = `
"<script lang=\\"ts\\" setup>
type _VTI_TYPE_Foo = string
interface Props {
  foo: _VTI_TYPE_Foo
}
defineProps<Props>()
</script>
"
`;

exports[`Common > Import export default > Use aliases > '3.ts' ('default') 1`] = `
"<script lang=\\"ts\\" setup>
type _VTI_TYPE_Foo = string
interface Props {
  foo: _VTI_TYPE_Foo
}
defineProps<Props>()
</script>
"
`;

exports[`Common > Import same type implicitly > Default > '1.ts' ('default') 1`] = `
"<script lang=\\"ts\\" setup>
type _VTI_TYPE_Foo = 'foo'
interface Props {
  foo: _VTI_TYPE_Foo
  bar: _VTI_TYPE_Foo
  baz: _VTI_TYPE_Foo
}
defineProps<Props>()
</script>
"
`;

exports[`Common > Import same type implicitly > Default > '2.ts' ('default') 1`] = `
"<script lang=\\"ts\\" setup>
type _VTI_TYPE_Foo = 'foo'
type _VTI_TYPE_A = _VTI_TYPE_Foo
interface Props {
  foo: _VTI_TYPE_Foo
  bar: _VTI_TYPE_Foo
  baz: _VTI_TYPE_A
}
defineProps<Props>()
</script>
"
`;

exports[`Common > Interface extends interface > Has reference > '1.ts' ('default') 1`] = `
"<script lang=\\"ts\\" setup>
type _VTI_TYPE_Baz = boolean
type _VTI_TYPE_Bar = number
type _VTI_TYPE_Foo = string
interface Props {
  baz: _VTI_TYPE_Baz
  foo: _VTI_TYPE_Foo
  bar: _VTI_TYPE_Bar
}
defineProps<Props>()
</script>
"
`;

exports[`Common > Interface extends interface > Has reference > '2.ts' ('default') 1`] = `
"<script lang=\\"ts\\" setup>
type _VTI_TYPE_Baz = boolean
interface _VTI_TYPE_BaseProps {
  baz: _VTI_TYPE_Baz
}
type _VTI_TYPE_Bar = number
type _VTI_TYPE_Foo = string
interface Props {
  baz: _VTI_TYPE_Baz
  foo: _VTI_TYPE_Foo
  bar: _VTI_TYPE_Bar
  base: _VTI_TYPE_BaseProps
}
defineProps<Props>()
</script>
"
`;

exports[`Common > Interface extends interface > Has reference > '3.ts' ('default') 1`] = `
"<script lang=\\"ts\\" setup>
type _VTI_TYPE_Qux = 'qux'
type _VTI_TYPE_Baz = boolean
interface _VTI_TYPE_BaseProps {
  qux: _VTI_TYPE_Qux
  baz: _VTI_TYPE_Baz
}
type _VTI_TYPE_Bar = number
type _VTI_TYPE_Foo = string
interface Props {
  qux: _VTI_TYPE_Qux
  baz: _VTI_TYPE_Baz
  foo: _VTI_TYPE_Foo
  bar: _VTI_TYPE_Bar
  base: _VTI_TYPE_BaseProps
}
defineProps<Props>()
</script>
"
`;

exports[`Common > Interface extends interface > No reference > 'index.ts' ('default') 1`] = `
"<script lang=\\"ts\\" setup>
interface Props {
  baz: boolean
  foo: string
  bar: number
}
defineProps<Props>()
</script>
"
`;

exports[`Common > Interface without reference > Default > 'index.ts' ('default') 1`] = `
"<script lang=\\"ts\\" setup>
interface Props {
  foo: string
  bar: number
}
defineProps<Props>()
</script>
"
`;

exports[`Common > Mixed aliases > Default > 'index.ts' ('default') 1`] = `
"<script lang=\\"ts\\" setup>
type _VTI_TYPE_F = 'A'
interface Props {
  foo: _VTI_TYPE_F
}
defineProps<Props>()
</script>
"
`;

exports[`Common > Multi level reference > Default > '1.ts' ('default') 1`] = `
"<script lang=\\"ts\\" setup>
type _VTI_TYPE_Foo = number
type _VTI_TYPE_Bar = _VTI_TYPE_Foo
interface Props {
  foo: _VTI_TYPE_Foo
  bar: _VTI_TYPE_Bar
}
defineProps<Props>()
</script>
"
`;

exports[`Common > Multi level reference > Default > '2.ts' ('default') 1`] = `
"<script lang=\\"ts\\" setup>
type _VTI_TYPE_Foo = number
type _VTI_TYPE_Bar = _VTI_TYPE_Foo
type _VTI_TYPE_Qux = _VTI_TYPE_Foo
type _VTI_TYPE_Baz = _VTI_TYPE_Bar
interface Props {
  foo: _VTI_TYPE_Foo
  bar: _VTI_TYPE_Bar
  baz: _VTI_TYPE_Baz
  qux: _VTI_TYPE_Qux
}
defineProps<Props>()
</script>
"
`;

exports[`Common > Redeclaration of types > Default > 'index.ts' ('default') 1`] = `
"<script lang=\\"ts\\" setup>
type _VTI_TYPE_Foo = [number, number]
interface Props {
  foo: _VTI_TYPE_Foo
  bar: _VTI_TYPE_Foo
}
defineProps<Props>()
</script>
"
`;

exports[`Common > Redeclaration of types > Same name > 'index.ts' ('default') 1`] = `
"<script lang=\\"ts\\" setup>
type _VTI_TYPE_Foo_2 = 'foo_2'
type _VTI_TYPE_Bar = _VTI_TYPE_Foo_2
type _VTI_TYPE_Foo = 'foo'
interface Props {
  foo: _VTI_TYPE_Foo
  bar: _VTI_TYPE_Bar
}
defineProps<Props>()
</script>
"
`;

exports[`Common > Reference in property > Default > 'index.ts' ('default') 1`] = `
"<script lang=\\"ts\\" setup>
interface _VTI_TYPE_Foo {
  foo: 'foo'
}
type _VTI_TYPE_Bar = _VTI_TYPE_Foo
interface Props {
  foo: _VTI_TYPE_Foo
  bar: _VTI_TYPE_Bar
}
defineProps<Props>()
</script>
"
`;

exports[`Common > Strict type finding > Default > 'index.ts' ('default') 1`] = `
"<script lang=\\"ts\\" setup>
interface Props {
  bar: string
}
defineProps<Props>()
</script>
"
`;
